cns/restserver/ipam.go

```diff
@@ -0,0 +1,280 @@
+// Copyright 2017 Microsoft. All rights reserved.
+// MIT License
+
+package restserver
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/Azure/azure-container-networking/cns"
+	"github.com/Azure/azure-container-networking/cns/logger"
+)
+
+func newIPConfig(ipAddress string, prefixLength uint8) cns.IPSubnet {
+	return cns.IPSubnet{
+		IPAddress:    ipAddress,
+		PrefixLength: prefixLength,
+	}
+}
+
+func NewPodState(ipaddress string, prefixLength uint8, id, ncid, state string) *cns.ContainerIPConfigState {
+	ipconfig := newIPConfig(ipaddress, prefixLength)
+
+	return &cns.ContainerIPConfigState{
+		IPConfig: ipconfig,
+		ID:       id,
+		NCID:     ncid,
+		State:    state,
+	}
+}
+
+func NewPodStateWithOrchestratorContext(ipaddress string, prefixLength uint8, id, ncid, state string, orchestratorContext cns.KubernetesPodInfo) (*cns.ContainerIPConfigState, error) {
+	ipconfig := newIPConfig(ipaddress, prefixLength)
+	b, err := json.Marshal(orchestratorContext)
+	return &cns.ContainerIPConfigState{
+		IPConfig:            ipconfig,
+		ID:                  id,
+		NCID:                ncid,
+		State:               state,
+		OrchestratorContext: b,
+	}, err
+}
+
+func (service *HTTPRestService) AddIPConfigsToState(ipconfigs []*cns.ContainerIPConfigState) error {
+	service.Lock()
+	defer service.Unlock()
+
+	for i, ipconfig := range ipconfigs {
+		service.PodIPConfigState[ipconfig.ID] = ipconfig
+		if ipconfig.State == cns.Allocated {
+			var podInfo cns.KubernetesPodInfo
+			err := json.Unmarshal(ipconfig.OrchestratorContext, &podInfo)
+			service.PodIPIDByOrchestratorContext[podInfo.GetOrchestratorContextKey()] = ipconfig.ID
+
+			// if batch request failed, remove added ipconfigs and return
+			if err != nil {
+				service.removeIPConfigsFromPool(ipconfigs[0:i])
+				return fmt.Errorf("Failed to add IPConfig to state: %+v with error: %v", ipconfig, err)
+			}
+		}
+	}
+	return nil
+}
+
+func (service *HTTPRestService) removeIPConfigsFromPool(ipconfigs []*cns.ContainerIPConfigState) error {
+	service.Lock()
+	defer service.Unlock()
+
+	for _, ipconfig := range ipconfigs {
+		service.PodIPConfigState[ipconfig.ID] = nil
+		var podInfo cns.KubernetesPodInfo
+		err := json.Unmarshal(ipconfig.OrchestratorContext, &podInfo)
+		if ipconfig.State == cns.Allocated {
+			service.PodIPIDByOrchestratorContext[podInfo.GetOrchestratorContextKey()] = ""
```

Comment:
do we also need set ip to available?

Reply:
This is called in the reduction rebatching scenario when the ipconfigs are removed from the CNS state. In the releaseIPConfigHandler the ipconfig as available but does not remove the ipconfig from the pool.

Reply:
Updated from offline